### PR TITLE
Add `ERC7674` (draft)

### LINKED ISSUE #4959: Transient Approval
In some cases, an ERC20 approval is granted, that is then used in the same transaction. In that case, using storage is expensive, and potentially dangerous if approval persists.

The idea is to have, next to the normal allowance, a transient allowance.

This transient allowance can be set using a new "approveTransient" function that replicates the behavior of "approve" (minus the event)

When doing a transferFrom (or any operation that consumes allowance) the transient allowance is used first. If that is enough, no storage read or right is necessary. If transient allowance is not enough, normal (storage allowance is used)

If EOA multicall becomes available, either through EIP-3074 or EIP-5806, EOAs will be able to:
- set transient allowance
- use that allowance in defi (swap) ?
In a single transaction. This should save significant amount of gas over the same approach with storage based allowances.

No allowance is left after the tx, reducing risks.

That does not require any change from the consumer side (same transferFrom).

[Implementation can be found here](https://github.com/Amxx/EIP-5806/blob/main/packages/test-uniswap/contracts/libs/ERC20TemporaryAllowance.sol)

### PR DESCRIPTION
**Pull request by Amxx**
Fixes #4959

[ERC-7674](https://github.com/ethereum/ERCs/pull/358) is still WIP, but having it implemented as a draft contract would be valuable for users

Pending discussions:
- [x] `if (value > 0) {...}` in `ERC20._spendAllowance` vs `if (value > 0) super._spendAllowance(...)` in `ERC20TemporaryApproval`
  - remove `_approve` call when value is 0
  - skip super call if temporary allowance covers everything
- [ ] emit event when temporary allowance is set/updated ?  

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changeset entry (run `npx changeset add`)

### CODE REVIEW

File: .changeset/serious-carrots-provide.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'openzeppelin-solidity': minor
+---
+
+`ERC20TemporaryAllowance`: add an ERC-20 extension that implements temporary approval using transient storage
```
Comment (Amxx) [inline .changeset/serious-carrots-provide.md:5]:
```suggestion
`ERC20TemporaryApproval`: add an ERC-20 extension that implements temporary approval using transient storage
```

File: test/token/ERC20/extensions/ERC20TemporaryApproval.test.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        amount: 50n,
+        temporaryExpected: ethers.MaxUint256,
+        persistentExpected: ethers.MaxUint256,
+      },
+    ]) {
+      persistentAllowance ??= 0n;
+      temporaryAllowance ??= 0n;
+      amount ??= 0n;
+      temporaryExpected ??= min(persistentAllowance + temporaryAllowance - amount, ethers.MaxUint256);
+      persistentExpected ??= persistentAllowance - max(0n, amount - temporaryAllowance);
```
Comment (Amxx) [inline test/token/ERC20/extensions/ERC20TemporaryApproval.test.js:80]:
```suggestion
      persistentExpected ??= persistentAllowance - max(amount - temporaryAllowance, 0n);
```

File: contracts/interfaces/draft-IERC7674.sol [hunk 3]
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+/**
+ * @dev Temporary Approval Extension for ERC-20 (https://github.com/ethereum/ERCs/pull/358[ERC-7674])
+ *
+ * WARNING: This ERC is not final, and is likely to evolve.
```
Comment (ernestognw) [inline contracts/interfaces/draft-IERC7674.sol:8]: I think this is implicit from [our docs](https://docs.openzeppelin.com/contracts/5.x/backwards-compatibility#draft_or_pre_final_ercs), and we haven't warned in other interfaces such as `draft-IERC1822.sol` or `draft-IERC6093`.

```suggestion
```

File: contracts/token/ERC20/ERC20.sol [hunk 4]
```diff
@@ -299,13 +299,15 @@ abstract contract ERC20 is Context, IERC20, IERC20Metadata, IERC20Errors {
...
+            uint256 currentAllowance = allowance(owner, spender);
+            if (currentAllowance != type(uint256).max) {
+                if (currentAllowance < value) {
+                    revert ERC20InsufficientAllowance(spender, currentAllowance, value);
+                }
+                unchecked {
+                    _approve(owner, spender, currentAllowance - value, false);
+                }
             }
         }
```
Comment (ernestognw) [inline contracts/token/ERC20/ERC20.sol:312]: Didn't we decide against [address poisoning](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3931) for similar reasons we would decide against filtering `value == 0`?

File: contracts/token/ERC20/README.adoc [hunk 5]
```diff
@@ -22,6 +22,7 @@ Additionally there are multiple custom extensions, including:
 * {ERC20FlashMint}: token level support for flash loans through the minting and burning of ephemeral tokens (standardized as ERC-3156).
 * {ERC20Votes}: support for voting and vote delegation.
 * {ERC20Wrapper}: wrapper to create an ERC-20 backed by another ERC-20, with deposit and withdraw methods. Useful in conjunction with {ERC20Votes}.
+* {ERC20TemporaryApproval}: support for temporary allowances following ERC-7674.
```
Comment (ernestognw) [inline contracts/token/ERC20/README.adoc:25]:
```suggestion
* {ERC20TemporaryApproval}: support for approvals lasting for only one transaction, as defined in ERC-7674
```

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ * WARNING: This is a draft contract. The corresponding ERC is still subject to changes.
+ */
+abstract contract ERC20TemporaryApproval is ERC20, IERC7674 {
+    using SlotDerivation for bytes32;
+    using StorageSlot for bytes32;
+    using StorageSlot for StorageSlot.Uint256SlotType;
+
+    // keccak256(abi.encode(uint256(keccak256("openzeppelin.storage.ERC20TemporaryApproval")) - 1)) & ~bytes32(uint256(0xff))
+    // solhint-disable-next-line const-name-snakecase
```
Comment (ernestognw) [inline contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:22]: Why not just having the constant in snake case?

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * @dev Alternative to {approve} that sets a `value` amount of tokens as the temporary allowance of `spender` over
+     * the caller's tokens.
+     *
+     * Returns a boolean value indicating whether the operation succeeded.
+     *
+     * Requirements:
+     * - `spender` cannot be the zero address.
+     *
+     * Does NOT emit an {Approval} event.
```
Comment (ernestognw) [inline contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:54]: This raises an interesting discussion about using events for signaling transient storage updates.
I first thought it shouldn't be needed, but it may disrupt indexers. The case I'm considering is when a company uses an indexer to report financial operations to authorities, for such cases there may be some transfer events they can't track because they were transient.

Whether it's relevant or not for most people is another discussion, but I see one of the main points of the ERC is to provide cheaper allowances, in that case, it'd be a matter of time before getting it widely adopted where possible.

What do you think?

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                // check how much of the value is covered by the transient allowance
+                uint256 spendTemporaryAllowance = Math.min(currentTemporaryAllowance, value);
+                // decrease transient allowance accordingly
+                _temporaryApprove(owner, spender, currentTemporaryAllowance - spendTemporaryAllowance);
+                // update value necessary
+                value -= spendTemporaryAllowance;
+            }
+            // reduce any remaining value from the persistent allowance
+            super._spendAllowance(owner, spender, value);
+        }
```
Comment (ernestognw) [inline contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:104]: This is safe, but it's wrapping a call to `allowance` in ERC20.sol
I'm worried that overriding `allowance` is not a crazy use case, and this extension would inadvertently uncheck its operations.

```suggestion
        }
        super._spendAllowance(owner, spender, value);
```

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+     * Does NOT emit an {Approval} event.
+     */
+    function _temporaryApprove(address owner, address spender, uint256 value) internal virtual {
+        if (owner == address(0)) {
+            revert ERC20InvalidApprover(address(0));
+        }
+        if (spender == address(0)) {
+            revert ERC20InvalidSpender(address(0));
+        }
+        ERC20TemporaryApprovalStorageLocation.deriveMapping(owner).deriveMapping(spender).asUint256().tstore(value);
```
Comment (cairoeth) [inline contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:80]: i think we can save a mload and keccak with something like:
```
  mstore(0x20, owner)
  mstore(0x0c, ERC20TemporaryApprovalStorageLocation)
  mstore(0x00, spender)
  tstore(keccak256(0x0c, 0x34), value)
```

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            // if there is temporary allowance
+            if (currentTemporaryAllowance > 0) {
+                // if infinite, do nothing
+                if (currentTemporaryAllowance == type(uint256).max) return;
+                // check how much of the value is covered by the transient allowance
+                uint256 spendTemporaryAllowance = Math.min(currentTemporaryAllowance, value);
+                // decrease transient allowance accordingly
+                _temporaryApprove(owner, spender, currentTemporaryAllowance - spendTemporaryAllowance);
+                // update value necessary
+                value -= spendTemporaryAllowance;
```
Comment (cairoeth) [inline contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:100]: value is only consumed once after this, so might as well just combine it in L103?

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 9 above)
Comment (cairoeth) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:80]: this would also work for `_temporaryAllowance`

File: contracts/token/ERC20/ERC20.sol (hunk 4 above)
Comment (Amxx) [reply contracts/token/ERC20/ERC20.sol:312]: The logic here is the following:
- if the temporary allowance is enough, we should not sload/sstore the persistent value (otherwize it break the point of gas savings). So we can do it in two ways:
  - in `ERC20TemporaryApproval._spendAllowance` only do the super call if value > 0
    - this is bad practice if someone else overrides `_spendAllowance`
  - in `ERC20` change the semantics of `_spendAllowance` to mean "if there is nothing to spend, we are good anyway".
  
If we look at `ERC20._spendAllowance`, this has the following impact
- if value = 0, `currentAllowance` cannot be smaller than value. `ERC20InsufficientAllowance` is never triggered, so the if doesn't change anything regarding the revert.
- if value is 0, 5.0 code does:
  - load the allowance (from a potentially overridable function)
  - substract zero from it
  - set it as the new allowance, without emitting an event.
  
So there is a change, we are no longer calling `_approve` with the current value. It may be possible to create edge cases where the missed call to an overrident `_approve` has an effect. IMO its a non issue.

Comment (Amxx) [reply contracts/token/ERC20/ERC20.sol:312]: About address poisoning, it doesn't chagne the possibility of doing it (or not doing it). It makes it cheaper though, because the poisonning call would not read/write the zero allowance.

Comment (Amxx) [reply contracts/token/ERC20/ERC20.sol:312]: @frangio curious to having your opinion on this if in the core ERC20.

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 7 above)
Comment (Amxx) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:54]: We already not have event to notify of allowance changes in transferFrom. Honestly, if we don't have those, I'm not sure why we would need one here.

On thing event are usefull for it tracking state changes ... but here, the (allowance) state doesn't really change. At the end of the tx its reset anyway. Its not like you are setting an allowance that will stay forever if you forger about it (that IMO needs to be notified)

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 8 above)
Comment (Amxx) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:104]: > I'm worried that overriding allowance is not a crazy use case, and this extension would inadvertently uncheck its operations.

Not sure I understand what you mean. You know that doing 
```
unchecked { fn1(a, b); }
``` 
does not propagate unchecked to chat happens inside `fn1` ?

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+     *
+     * Does NOT emit an {Approval} event.
+     */
+    function _temporaryApprove(address owner, address spender, uint256 value) internal virtual {
+        if (owner == address(0)) {
+            revert ERC20InvalidApprover(address(0));
+        }
+        if (spender == address(0)) {
+            revert ERC20InvalidSpender(address(0));
+        }
```
[the remark this answers is not in the document]
Comment (Amxx) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:79]: I'm duplicating the logic we have in `ERC20._approve`.

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 9 above)
Comment (Amxx) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:80]: What I'm proposing matches the derivation of solidity. That means in the future, it solidity supports it, we can move to 
```solidity
struct ERC20TemporaryApprovalStorage {
    mapping(address account => mapping(address spender => uint256)) private _tAllowances;
}

function _getERC20TemporaryApprovalStorage() private pure returns (ERC20TemporaryApprovalStorage transient $) {
    assembly {
        $.slot := ERC20TemporaryApprovalStorageLocation
    }
}
```
and that would keep using the same locations

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 10 above)
Comment (Amxx) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:100]: `spendTemporaryAllowance` is only defined inside the if. what you propose would require moving the declaration of `spendTemporaryAllowance` oustide the if, and doing the `-` operation all the time, even when `currentTemporaryAllowance` is `0`, and we know its not necessary. 

We can try, but I believe it will make the code more expensive

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 6 above)
Comment (Amxx) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:22]: We cannot declare transient state, so we cannot do stuff the "normal" way (and have the transpiler do the transformation) ... so I though of manually implementing the "upgradeable" way.

Like we do in Initializable.sol

I though it would be simpler to just use the same syntax as the upgradeable contract does, but we can change that.

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 9 above)
Comment (Amxx) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol]: Also:
- the `deriveMapping` function use the scratch space. They do not expand memory (cost is ~50gas each time)
- doing a single hash operation will require having all 3 values next to each other. both addresses are 20 bytes, to have that fit into the scratch space, we would need the "prefix" (mapping location equivalent) to be 24 bytes and not 32 as usual. If we stick with the usual 32 bytes, we cannot use the scratch space and are potentially extending memory, which has a cost.

File: contracts/token/ERC20/ERC20.sol (hunk 4 above)
Comment (frangio) [reply contracts/token/ERC20/ERC20.sol:312]: > It may be possible to create edge cases where the missed call to an overrident `_approve` has an effect.

Hm, I have a vague memory that this actually was a concern for a project once... It does seem quite risky to change this in a minor version.

Comment (frangio) [reply contracts/token/ERC20/ERC20.sol:312]: Could we implement this by overriding only `allowance` and `_approve` instead of `_spendAllowance`?

Comment (Amxx) [reply contracts/token/ERC20/ERC20.sol:312]: > Could we implement this by overriding only allowance and _approve instead of _spendAllowance?

I have no idea how. In particular I'm not sure how to make transferFrom spend only temporary allowance, without touching the normal allowance, when the temporary allowance is enough

Comment (Amxx) [reply contracts/token/ERC20/ERC20.sol:312]: Discussion on the ERC wording here: https://github.com/ethereum/ERCs/pull/358/commits/3c10621486d401040e187e56c31e570f292926c4#r1633595540

Comment (frangio) [reply contracts/token/ERC20/ERC20.sol:312]: I think it's possible, but not without the same problem of not invoking `super._approve`. There seems to be no way around that.

Comment (frangio) [reply contracts/token/ERC20/ERC20.sol:312]: > - in `ERC20TemporaryApproval._spendAllowance` only do the super call if value > 0
> 
>   -  this is bad practice if someone else overrides `_spendAllowance`

This is true, but it may be preferable than changing the behavior in a minor version. I think it's worth considering.

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 7 above)
Comment (frangio) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:53]: I've thought about this in the past [for ERC-6909X](https://github.com/frangio/ERCs/blob/erc6909x/ERCS/eip-xxxx.md#approval-and-operatorset-events) (a proposed extension I designed) and ended up deciding to make events a "SHOULD":

>ERC-6909 requires `Approval` and `OperatorSet` events to be emitted when allowance and operator status respectively are set. ERC-6909X tokens SHOULD emit these events as part of temporary approvals for strict compliance. The omission of these events during temporary approvals may confuse indexers that rely on events to track allowances and operators. For example, an indexer that assumes the events are complete may conclude that a spender has zero allowance in a case where in fact it has non-zero allowance.

It's not a "MUST" because I understand that for max savings tokens would want to omit them.

But that was in the context of ERC-6909 compliance...



> We already not have event to notify of allowance changes in transferFrom. Honestly, if we don't have those, I'm not sure why we would need one here.

This is a valid point.

File: contracts/token/ERC20/ERC20.sol (hunk 4 above)
Comment (Amxx) [reply contracts/token/ERC20/ERC20.sol:312]: Its what I used to do, before realizing there was a way to have the super call always happen.
I'm leanning toward the current version, but I'm open to re-using the old one

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 7 above)
Comment (ernestognw) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:53]: > IMO not having events would make it harder to look for historical activity/usage 

Right. Based on other's comments I realize it's not a big deal since users may decide whether or not to add an event. That's the decision we took by removing the event from ERC20's `_spendAllowance`.

> We already not have event to notify of allowance changes in transferFrom. Honestly, if we don't have those, I'm not sure why we would need one here.

Yes, I agree with this. The difference is that the event is not defined in this case whereas regular ERC20 users would identify the lack of the `Approval` event during testing and before deploying/upgrading.

Seems like this is a recommendation in the ERC already. It should be specific: https://github.com/ethereum/ERCs/pull/358/files#r1633892313

File: contracts/token/ERC20/ERC20.sol (hunk 4 above)
Comment (Amxx) [reply contracts/token/ERC20/ERC20.sol]: As discussed, I re-implemented the old one.

It negativelly affect the tests because the error emitted in some edge conditions depends on _approve being called (with spender = 0 and value = 0).

File: test/token/ERC20/ERC20.behavior.js [hunk 12]
```diff
@@ -132,9 +132,18 @@ function shouldBehaveLikeERC20(initialSupply, opts = {}) {
     });
 
     it('reverts when the token owner is the zero address', async function () {
+      // transferFrom does a spendAllowance before moving the assets
+      // - default behavior (ERC20) is to always update the approval using `_approve`. This will fail because the
+      //   approver (owner) is address(0). This happens even if the amount transfered is zero, and the approval update
```
Comment (Amxx) [inline test/token/ERC20/ERC20.behavior.js:137]:
```suggestion
      //   approver (owner) is address(0). This happens even if the amount transferred is zero, and the approval update
```

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 8 above)
Comment (ernestognw) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:108]: I didn't know! I was always manually make sure propagated `unchecked` blocks were not an issue.
Solving

File: contracts/interfaces/draft-IERC7674.sol [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+pragma solidity ^0.8.20;
+
+/**
+ * @dev Temporary Approval Extension for ERC-20 (https://github.com/ethereum/ERCs/pull/358[ERC-7674])
+ */
+interface IERC7674 {
+    /**
+     * @dev Set the temporary allowance, allowing allows `spender` to withdraw (within the same transaction) assets
+     * held by the caller.
```
Comment (cairoeth) [inline contracts/interfaces/draft-IERC7674.sol:11]: https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5071/files#diff-7b0f1795668241cd100ce0a5728fc598e10533cf7bd2846a8b5c86b8a7ac6fa5R60-R71 is much clearer imo. for consistency (and to not duplicate), shouldn't we have these comments in the interface, and point to the interface in the implementation function comments? as in here https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L118

File: contracts/interfaces/draft-IERC7674.sol [hunk 14]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+/**
+ * @dev Temporary Approval Extension for ERC-20 (https://github.com/ethereum/ERCs/pull/358[ERC-7674])
+ */
+interface IERC7674 {
+    /**
+     * @dev Set the temporary allowance, allowing allows `spender` to withdraw (within the same transaction) assets
```
Comment (cairoeth) [inline contracts/interfaces/draft-IERC7674.sol:10]:
```suggestion
     * @dev Set the temporary allowance, allowing `spender` to withdraw (within the same transaction) assets
```

**Review by cairoeth — COMMENTED** [review]
implementation lgtm. codecov is failing, rebasing should fix.

File: contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol (hunk 7 above)
Comment (Amxx) [reply contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol:53]: We agreed to not have an event during today's call. Solving

**Review by cairoeth — APPROVED** [review]
lgtm!

**Review by ernestognw — APPROVED** [review]
Looking good for a draft. Thanks all!
